Implement special backoff behavior on HTTP 409

bitglue opened the pull request:
It's clear from the documentation that Google may respond with a message
asking the client to back off, but it's not clear what form that takes
over the HTTP API. It seems though it could be an HTTP 409, so make an
effort to parse and respect that.

Since the looping and retry logic was getting complicated enough, it's
been extracted to a separate Looper class. Besides the new behavior on
HTTP 4xx responses, it should be unchanged.

bitglue committed:
Implement special backoff behavior on HTTP 409

It's clear from the documentation that Google may respond with a message
asking the client to back off, but it's not clear what form that takes
over the HTTP API. It seems though it could be an HTTP 409, so make an
effort to parse and respect that.

Since the looping and retry logic was getting complicated enough, it's
been extracted to a separate Looper class. Besides the new behavior on
HTTP 4xx responses, it should be unchanged.

arlandism commented on lib/cloud_profiler_agent/looper.rb:
```diff
@@ -0,0 +1,81 @@
+  # appropriate.
+  class Looper
+    def initialize(
+      min_iteration_time: 10,
```
The unit would be nice for the times, either as a comment or suffix.

EDIT: I do see the comment on line 19. Still wonder if maybe it'd be nice from the caller's perspective.

bitglue replied:
good point, changed these to `min_iteration_sec`, etc

bitglue committed:
Explicitly set interval for Stackprof

1000 is the default for :cpu and :wall, but the default for :object is
1. That would probably lead to very large :object profiles, so let's
explicitly set 1000 always.

arlandism commented on lib/cloud_profiler_agent/looper.rb:
```diff
@@ -0,0 +1,91 @@
+            iteration_time = @max_iteration_time
+          else
+            debug_log("sleeping for #{backoff} at request of server")
+            @sleeper.call(backoff)
```
So max iteration time can be exceeded if Google asks us to back off

bitglue replied:
yes, that's intentional. Google throttles clients to get 1 profile per minute, so with a very large number of clients they might ask us to back off for a pretty long time.

arlandism commented on spec/looper_spec.rb:
```diff
@@ -0,0 +1,157 @@
+    end
+
+    it 'runs the block forever when max_iterations is not given' do
+      # but we don't test this, because how do you test an infinite loop?
```
Agreed re: infinite loops. In those scenarios, it seems more important to just test what happens in one tick of the loop vs the loop construct itself. If you really were looking for a way to test you could DI another abstraction that decides when the loop is finished, mock it here, and then test the concrete abstraction logic in another file. TBH, that seems a little more trouble than it's worth here but is an option.

bitglue reviewed (commented):
just force pushed to fix an undefined `e` when StandardError is rescued.

arlandism commented on spec/looper_spec.rb:
```diff
@@ -0,0 +1,157 @@
+    context 'when the block takes some time' do
+      it 'accounts for time taken by the block' do
+        subject.run(3) { @now += 1 }
+        expect(@sleeps).to eq([min_time - 1, min_time - 1])
```
clever
